tfsdk/debug.go

```diff
@@ -0,0 +1,164 @@
+	// this manner, Terraform CLI disables certain plugin handshake checks and
+	// will not stop the provider process.
+	envTfReattachProviders = "TF_REATTACH_PROVIDERS"
+)
+
+// reattachConfig holds the information Terraform needs to be able to attach
+// itself to a provider process, so it can drive the process.
+//
+// This duplicate implementation is required because plugin.ReattachConfig.Addr
+// is not friendly for JSON encoding.
+type reattachConfig struct {
+	Protocol        string
+	ProtocolVersion int
+	Pid             int
+	Test            bool
+	Addr            reattachConfigAddr
+}
+
+// reattachConfigAddr is a JSON-encoding friendly version of net.Addr.
+type reattachConfigAddr struct {
+	Network string
+	String  string
+}
+
+// debugServe starts the plugin server in debug mode. This should only be used
+// when the provider will manage its own lifecycle. It is not recommended for
+// normal usage; Serve is the correct function for that. Debug can be used
+// in provider main functions for debugging processes, such as delve.
+//
+// This function is currently unexported unless there is a use case for
+// directly orchestrating the provider, e.g. via terraform-exec.
+func debugServe(ctx context.Context, providerFunc func() Provider, opts ServeOpts) (reattachConfig, <-chan struct{}, error) {
+	closeCh := make(chan struct{})
+	errCh := make(chan error)
+	pluginReattachConfigCh := make(chan *plugin.ReattachConfig)
+	tf6ProviderServerFunc := func() tfprotov6.ProviderServer {
+		return &server{
+			p: providerFunc(),
+		}
+	}
+	tf6ServerOpts := []tf6server.ServeOpt{
+		tf6server.WithDebug(ctx, pluginReattachConfigCh, closeCh),
+	}
+
+	go func() {
+		errCh <- tf6server.Serve(opts.Name, tf6ProviderServerFunc, tf6ServerOpts...)
+	}()
+
+	var pluginReattachConfig *plugin.ReattachConfig
+
+	select {
+	case err := <-errCh:
+		return reattachConfig{}, closeCh, err
+	case pluginReattachConfig = <-pluginReattachConfigCh:
+	case <-time.After(defaultReattachConfigTimeout):
+		return reattachConfig{}, closeCh, errors.New("timeout waiting on reattach config")
+	}
+
+	if pluginReattachConfig == nil {
+		return reattachConfig{}, closeCh, errors.New("nil reattach config received")
+	}
+
+	return reattachConfig{
+		Protocol:        string(pluginReattachConfig.Protocol),
+		ProtocolVersion: pluginReattachConfig.ProtocolVersion,
+		Pid:             pluginReattachConfig.Pid,
+		Test:            pluginReattachConfig.Test,
+		Addr: reattachConfigAddr{
+			Network: pluginReattachConfig.Addr.Network(),
+			String:  pluginReattachConfig.Addr.String(),
+		},
+	}, closeCh, nil
+}
+
+// Debug runs the provider in a mode acceptable for debugging and testing
+// processes, such as delve, by managing the process lifecycle. Information
+// needed for Terraform CLI to connect to the provider is output to stdout.
+// os.Interrupt (Ctrl-c) can be used to stop the provider.
+func Debug(ctx context.Context, providerFunc func() Provider, opts ServeOpts) error {
```

Comment:
Aside: Instead of introducing an exported function, we could also implement this by adding a `Debug bool` field to `ServeOpts`, then triggering this logic via an unexported function. Thinking about that again, maybe that would be a cleaner implementation for providers rather than forcing them to implement the conditional between serving functions.

Reply:
Opted to go with the `ServeOpts` approach!